client/src/main/java/io/pravega/client/stream/impl/TxnKeepAlive.java

```diff
@@ -0,0 +1,114 @@
+    private final AtomicBoolean isStarted = new AtomicBoolean();
+    @GuardedBy("$lock")
+    private ScheduledFuture<?> pingTxnTask;
+
+    TxnKeepAlive(EventWriterConfig config, Stream stream, Controller controller, ScheduledExecutorService
+            internalExecutor) {
+        this.txnLeaseMillis = config.getTransactionTimeoutTime();
+        this.pingIntervalMillis = getKeepAliveInterval(txnLeaseMillis);
+        this.stream = stream;
+        this.controller = controller;
+        this.executor = internalExecutor;
+    }
+
+    void startTxnKeepAlive(UUID txnID) {
+        txnList.add(txnID);
+        startPeriodicPingTxn();
+    }
+
+    void stopTxnKeepAlive(UUID txnID) {
+        txnList.remove(txnID);
+    }
+
+    private long getKeepAliveInterval(long txnLeaseMillis) {
+        double keepAliveInterval = txnLeaseMillis * KEEP_ALIVE_INTERVAL_FACTOR;
+        if (keepAliveInterval < TimeUnit.SECONDS.toMillis(10)) {
+            log.warn("Transaction Keep Alive interval is less than 10 seconds(lower bound)");
+        }
+        //Ping interval cannot be less than KeepAlive task interval of 10seconds.
+        return Math.max(TimeUnit.SECONDS.toMillis(10), (long) keepAliveInterval);
+    }
+
+    @Synchronized
+    private void startPeriodicPingTxn() {
+        if (!isStarted.get()) {
+            log.info("Starting Transaction keep alive at an interval of {}ms ", this.pingIntervalMillis);
+            pingTxnTask = executor.scheduleAtFixedRate(this::pingTransactions, 10, this.pingIntervalMillis,
+                    TimeUnit.MILLISECONDS);
+            isStarted.set(true);
+        }
+    }
+
+    /*
+        Ping all the transactions present in the list. Controller client performs retries in case of a failures. On a
+         failure the particular transaction it is removed from the ping list.
+     */
+    private void pingTransactions() {
+        Map<UUID, CompletableFuture<Void>> pingResult =
+                txnList.stream().collect(Collectors.toMap(Function.identity(),
+                        uuid -> controller.pingTransaction(stream, uuid, txnLeaseMillis)));
+        Futures.allOf(pingResult.values())
+               .whenComplete((v, ex) -> pingResult.entrySet().stream()
+                                                  .filter(e -> !Futures.isSuccessful(e.getValue()))
+                                                  .forEach(e -> {
+                                                      log.warn("Ping Transaction for txn ID:{} failed", e.getKey(),
+                                                              getException(e.getValue()));
+                                                      txnList.remove(e.getKey()); //ping txn failed, remove it.
+                                                  }));
+    }
+
+    @Override
+    public void close() {
+        log.info("Closing Transaction keep alive periodic task");
+        if (pingTxnTask != null) {
```

Comment:
pingTxnTasks is not accessed under the lock.
You need to extract it to a local variable, set it to null, and then outside of the synchronized block you need to call `cancel` (if it's not null)

Reply:
done.